### PR TITLE
Fix #13: Support textobjects defined by start/end node

### LINKED ISSUE #13: Support textobjects defined by start and end node
#12 

We could define texobjects like in #12 with by their start and end nodes. E.g.:
```
(
  (method_signature) @function.outer.start
  (function_body) @function.outer.end
) 
```
or in Lua
```
(local_function . (_) . (_) . (_) @function.inner.start  (_) @function.inner.end .)
```

### PR DESCRIPTION
**Pull request by theHamsta**
Requires https://github.com/nvim-treesitter/nvim-treesitter/pull/553

This does not fix the "requires refactor" problem of that function.

### CODE REVIEW

File: lua/nvim-treesitter/textobjects/shared.lua [hunk 1]
```diff
@@ -7,6 +7,31 @@ local ts_utils = require'nvim-treesitter.ts_utils'
...
+  if not node2 then return end
+  local start_line, start_col, end_line, end_col
+  if node1 and node2 then
+   start_line, start_col = node1:start()
+   end_line, end_col = node2:end_()
+  else
+   start_line, start_col, end_line, end_col = node2:range()
+  end
+
+  if line >= start_line and line <= end_line then
```
Comment (theHamsta) [inline lua/nvim-treesitter/textobjects/shared.lua:20]: should upstream that `ts.is_in_node_range` could also support ranges.

File: lua/nvim-treesitter/textobjects/shared.lua [hunk 2]
```diff
@@ -40,8 +65,8 @@ function M.textobject_at_point(query_string, pos, bufnr)
   local earliest_start
 
   for _, m in pairs(matches) do
-    if m.node and ts_utils.is_in_node_range(m.node, row, col) then
-      local length = ts_utils.node_length(m.node)
+    if is_in_node_range(m.start and m.start.node, m.node or (m['end'] and m['end'].node), row, col) then
+      local length = ts_utils.node_length(m['end'] and m['end'].node or m.node)
```
Comment (theHamsta) [inline lua/nvim-treesitter/textobjects/shared.lua:69]: TODO: length should end-start in the start end case.

**Comment by theHamsta** [comment]
Maybe we can incorporate the refactor and also select to next textobject if cursor is between textobject.

**Comment by theHamsta** [comment]
I guess it's better when we would have a (make-range! "function.outer" function.outer.start function.outer.end) in nvim-treesitter instead. Then, this feature would not require a code change and other plugins could create virtual nodes by start and end, too.

**Comment by GenesisTMS** [comment]
> I guess it's better when we would have a `(make-range! "function.outer" function.outer.start function.outer.end)` in nvim-treesitter instead.

I am voting for vanila approach with sequence.
ofc i don't know how much work it will cost but with basic sequence you can describe any scenario (or am i wrong?)

```
(
  (method_signature)
  (function_body)
) @function.outer
```

or am i missed the point? :D trying to catch up

**Comment by theHamsta** [comment]
@GenesisTMS sorry, I posted to the wrong issue. My solution for this problem involves to much code changes and it would be better to express those node ranges in the queries.

Btw.

```scheme
((
  (method_signature) @start
   .  ; the dot is necessary since you don't want to match with the signature of *another*  function `.` means the nodes are adjacent
  (function_body) @end
) 
(make-range! "function.outer" @start @stop)) ; create a new virtual node by the range defined by start/end
```

**Comment by theHamsta** [comment]
Superseded by #16

**Pull request closed without merging**